File: iosApp/iosApp/Modules/CurrencyPicker/CurrencyPickerWireframe.swift
// Created by web3d4v on 06/06/2022.
// Copyright (c) 2022 Sons Of Crypto.
// SPDX-License-Identifier: MIT

import UIKit
import web3walletcore

final class DefaultCurrencyPickerWireframe {
    
    private weak var parent: UIViewController?
    private let context: CurrencyPickerWireframeContext
    private let currencyAddWireframeFactory: CurrencyAddWireframeFactory
    private let walletService: WalletService
    private let networksService: NetworksService
    private let currencyStoreService: CurrencyStoreService
    
    private weak var vc: UIViewController?
    
    init(
        _ parent: UIViewController?,
        context: CurrencyPickerWireframeContext,
        currencyAddWireframeFactory: CurrencyAddWireframeFactory,
        walletService: WalletService,
        networksService: NetworksService,
        currencyStoreService: CurrencyStoreService
    ) {
        self.parent = parent
        self.context = context
        self.currencyAddWireframeFactory = currencyAddWireframeFactory
        self.walletService = walletService
        self.networksService = networksService
        self.currencyStoreService = currencyStoreService
    }
}

extension DefaultCurrencyPickerWireframe: CurrencyPickerWireframe {
    
    func present() {
        let vc = wireUp()
        parent?.present(vc, animated: true)
    }
    
    func navigate(destination___ destination: CurrencyPickerWireframeDestination) {
        if let target = destination as? CurrencyPickerWireframeDestination.AddCustomCurrency {
            currencyAddWireframeFactory.make(
                vc,
                context: .init(network: target.network)
            ).present()
        }
        if (destination as? CurrencyPickerWireframeDestination.Dismiss) != nil {
            vc?.popOrDismiss()
        }
    }
}

private extension DefaultCurrencyPickerWireframe {
    
    func wireUp() -> UIViewController {
        let interactor = DefaultCurrencyPickerInteractor(
            walletService: walletService,
            currencyStoreService: currencyStoreService
        )
        let vc: CurrencyPickerViewController = UIStoryboard(.currencyPicker).instantiate()
        let presenter = DefaultCurrencyPickerPresenter(
            view: WeakRef(referred: vc),
            wireframe: self,
            interactor: interactor,
            context: context
        )
        vc.presenter = presenter
        vc.context = context
        let nc = NavigationController(rootViewController: vc)
        self.vc = nc
        return nc
    }
}

File: androidApp/src/main/java/com/sonsofcrypto/web3wallet/android/modules/compose/currencypicker/CurrencyPickerWireframe.kt
package com.sonsofcrypto.web3wallet.android.modules.compose.currencypicker

import androidx.fragment.app.Fragment
import com.sonsofcrypto.web3lib.services.currencyStore.CurrencyStoreService
import com.sonsofcrypto.web3lib.services.wallet.WalletService
import com.sonsofcrypto.web3lib.utils.WeakRef
import com.sonsofcrypto.web3wallet.android.common.extensions.navigationFragment
import com.sonsofcrypto.web3wallet.android.common.ui.navigationFragment
import com.sonsofcrypto.web3wallet.android.modules.compose.currencyadd.CurrencyAddWireframeFactory
import com.sonsofcrypto.web3walletcore.modules.currencyAdd.CurrencyAddWireframeContext
import com.sonsofcrypto.web3walletcore.modules.currencyPicker.*

class DefaultCurrencyPickerWireframe(
    private val parent: WeakRef<Fragment>?,
    private val context: CurrencyPickerWireframeContext,
    private val walletService: WalletService,
    private val currencyStoreService: CurrencyStoreService,
    private val currencyAddWireframeFactory: CurrencyAddWireframeFactory,
): CurrencyPickerWireframe {

    override fun present() {
        val fragment = wireUp()
        parent?.navigationFragment?.present(fragment, animated = true)
    }

    override fun navigate(destination: CurrencyPickerWireframeDestination) {
        when (destination) {
            is CurrencyPickerWireframeDestination.AddCustomCurrency -> {
                val context = CurrencyAddWireframeContext(destination.network)
                currencyAddWireframeFactory.make(parent?.get(), context).present()
            }
            is CurrencyPickerWireframeDestination.Dismiss -> {
                parent?.navigationFragment?.popOrDismiss()
            }
        }
    }

    private fun wireUp(): Fragment {
        val view = CurrencyPickerFragment()
        val interactor = DefaultCurrencyPickerInteractor(
            walletService, currencyStoreService
        )
        val presenter = DefaultCurrencyPickerPresenter(
            WeakRef(view),
            this,
            interactor,
            context
        )
        view.presenter = presenter
        return view
    }
}